src/openrct2/world/Duck.cpp

```diff
@@ -0,0 +1,398 @@
+    6, 6, 6, 5, 5, 5, 5, 4, 4, 4, 4, 0, 0, 0, 0, 0xFF
+};
+
+static const uint8 DuckAnimationFlyAway[] =
+{
+    8, 9, 10, 11, 12, 13
+};
+
+static const uint8 * DuckAnimations[] =
+{
+    DuckAnimationFlyToWater,    // FLY_TO_WATER
+    DuckAnimationSwim,          // SWIM
+    DuckAnimationDrink,         // DRINK
+    DuckAnimationDoubleDrink,   // DOUBLE_DRINK
+    DuckAnimationFlyAway,       // FLY_AWAY
+};
+
+bool rct_sprite::IsDuck()
+{
+    return this->duck.sprite_identifier == SPRITE_IDENTIFIER_MISC &&
+           this->duck.misc_identifier == SPRITE_MISC_DUCK;
+}
+
+rct_duck * rct_sprite::AsDuck()
+{
+    rct_duck * result = nullptr;
+    if (IsDuck())
+    {
+        return (rct_duck *)result;
+    }
+    return result;
+}
+
+void rct_duck::Invalidate()
+{
+    invalidate_sprite_0((rct_sprite *)this);
+}
+
+void rct_duck::Remove()
+{
+    sprite_remove((rct_sprite *)this);
+}
+
+void rct_duck::MoveTo(sint16 destX, sint16 destY, sint16 destZ)
+{
+    sprite_move(destX, destY, destZ, (rct_sprite *)this);
+}
+
+void rct_duck::UpdateFlyToWater()
+{
+    if ((gCurrentTicks & 3) != 0) return;
+
+    frame++;
+    if (frame >= Util::CountOf(DuckAnimationFlyToWater))
+    {
+        frame = 0;
+    }
+
+    Invalidate();
+    sint32 manhattanDistance = abs(target_x - x) + abs(target_y - y);
+    sint32 direction = sprite_direction >> 3;
+    sint32 newX = x + DuckMoveOffset[direction].x;
+    sint32 newY = y + DuckMoveOffset[direction].y;
+    sint32 manhattanDistanceN = abs(target_x - newX) + abs(target_y - newY);
+
+    rct_map_element * mapElement = map_get_surface_element_at(target_x >> 5, target_y >> 5);
+    sint32 waterHeight = mapElement->properties.surface.terrain & 0x1F;
+    if (waterHeight == 0)
+    {
+        state = DUCK_STATE::FLY_AWAY;
+        UpdateFlyAway();
+    }
+    else
+    {
+        waterHeight <<= 4;
+        sint32 newZ = abs(z - waterHeight);
+
+        if (manhattanDistanceN <= manhattanDistance)
+        {
+            if (newZ > manhattanDistanceN)
+            {
+                newZ = z - 2;
+                if (waterHeight >= z)
+                {
+                    newZ += 4;
+                }
+                frame = 1;
+            }
+            else
+            {
+                newZ = z;
+            }
+            MoveTo(newX, newY, newZ);
+            Invalidate();
+        }
+        else
+        {
+            if (newZ > 4)
+            {
+                state = DUCK_STATE::FLY_AWAY;
+                UpdateFlyAway();
+            }
+            else
+            {
+                state = DUCK_STATE::SWIM;
+                frame = 0;
+                UpdateSwim();
+            }
+        }
+    }
+}
+
+void rct_duck::UpdateSwim()
+{
+    if (((gCurrentTicks + sprite_index) & 3) != 0) return;
+
+    uint32 randomNumber = scenario_rand();
+    if ((randomNumber & 0xFFFF) < 0x666)
+    {
+        if (randomNumber & 0x80000000)
```

Comment:
In some of the peep code I would just convert this into a new `scenario_rand` call. I'm not fussed if you think its better to do it this way though.